Initial vhost-user-scsi implementation

Gaelan opened the pull request:
Here's an initial implementation of vhost-user-scsi, as part of my GSoC project with @slp. Currently, it supports reading from up to 256 raw images.

For usage documentation and plans for future improvements, see [the README](https://github.com/rust-vmm/vhost-device/blob/da92bee292e1f4ba8683ee1288adeb0c67c04a65/src/scsi/README.md); for architectural documentation, see [ARCHITECTURE.md](https://github.com/rust-vmm/vhost-device/blob/da92bee292e1f4ba8683ee1288adeb0c67c04a65/src/scsi/ARCHITECTURE.md) and comments in the code.

Although a read-only disk driver is of limited usefulness, I believe this code should be more-or-less production-ready. I've left some TODO comments in the code; these are either for features we'll want to add in the future, or places I want feedback from code reviewers (especially who are familiar with SCSI).

Gaelan commented:
Looks like the CI is failing, for a strange reason: The tests depend on an image file, which they finds relative to the current directory. This works locally and in the "unittests" phase of CI, but apparently not when collecting coverage data - I guess that infrastructure somehow manages to run tests from a different directory. I'll investigate more carefully tomorrow.

Gaelan commented:
Oh, I forgot: this currently depends on @slp's fork of `vhost-user-backend`; we'll probably want to merge rust-vmm/vhost-user-backend#18 before we merge this.

Gaelan commented:
Just pushed a test refactor, back ported from my work on write support, that should fix the CI support.

Gaelan commented:
Ok, well CI still doesn't pass, but now it's for good reasons (a clippy warning I forgot to fix, and needing to change the coverage threshold).

vireshk commented:
Hi @Gaelan Can you rebase this stuff and fix the conflicts ? I will try to review it then. thanks.

Gaelan commented:
Just rebased, but it looks like something isn't compiling anymore - I think there may have been breaking changes in vhost-user-backend? I'll take a look.

Gaelan commented:
Wait, it's vhost-user-backend itself that doesn't compile. Hmm.

vireshk commented:
> Wait, it's vhost-user-backend itself that doesn't compile. Hmm.

You are still using an older version of vhost and backend. Please look at i2c's Cargo.toml. And I am not sure if we should be locking fixed version of  Cargo.lock or just keep moving to the latest stuff. I haven't added the same for i2c, maybe you should merge it as well ?

jiangliu commented:
Sorry for this, we are refactoring the vhost-user-backend crate. And there's still coming break changes:(

vireshk commented:
> Sorry for this, we are refactoring the vhost-user-backend crate. And there's still coming break changes:(

I don't see any failures on master though (#f57859ea). Perhaps once Gaelan moves to master, this will get fixed.

vireshk commented:
> I don't see any failures on master though (#f57859ea). Perhaps once Gaelan moves to master, this will get fixed.

My bad. I realized it later that Gaelan is basing stuff on new backend stuff.

Gaelan commented:
Alright, I've done a rebase and fixed some breaking changes. I should note that I've run the tests and they pass, but I haven't actually tried to run a real VM with the rebased code—I should have some free time in the next few days to give that a shot.

Sorry for the long delay—I've been busy with school.

stsquad reviewed (commented):
This test image seems like something we should generate on the developer machine rather than host in the GIT repo.

stsquad commented on src/scsi/.gitignore:
```diff
@@ -0,0 +1,2 @@
+/target
+target
```
is this a duplicate? Are we trying to ignore a file called target as well as the build directory?

Gaelan replied:
Not sure what I was thinking there. In any case, this gitignore seems to be an artifact from before I moved the crate into the `vhost-device` repo, and it doesn't do anything now, so I'll remove it.

stsquad commented on src/scsi/src/scsi/tests/test.img:
```diff
@@ -0,0 +1 @@
+000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 ...
```
We should generate this file in the build rather than have it in CI

Gaelan replied:
Good idea, done.

stsquad commented on src/scsi/README.md:
````diff
@@ -0,0 +1,43 @@
+
+```
+vhost-user-scsi -r /tmp/vhost-user-scsi.sock /path/to/image.raw /path/to/second-image.raw ...
+```
````
With my testing it crapped out on me:

    ➜  ./target/debug/vhost-user-scsi -r /tmp/vhost-user-scsi.sock test.img
    [2021-12-07T11:46:48Z ERROR vhost_user_scsi] Error running daemon: HandleRequest(InvalidParam)
    [src/scsi/src/main.rs:391]

Gaelan replied:
Hmm, not easily reproducing this (AlmaLinux 8.5 host, self-built QEMU 6.2, Debian 11 "nocloud" qcow2 image as guest). Some thoughts:

* Setting `RUST_LOG=debug` will get you a better idea of when things go wrong.
* Does this happen as soon as you start `vhost-user-scsi`, or after QEMU connects?
* What QEMU version and guest OS are you running?

stsquad commented on src/scsi/src/scsi/tests/mod.rs:
```diff
@@ -0,0 +1,362 @@
+    let mut f = tempfile().unwrap();
+    f.write_all(include_bytes!("./test.img")).unwrap();
+    f
+}
```
See previous comment about including test image in GIT repo.

stsquad commented on src/scsi/src/main.rs:
```diff
@@ -0,0 +1,392 @@
+    fn features(&self) -> u64 {
+        // TODO: Any other ones worth implementing? EVENT_IDX and INDIRECT_DESC
+        // are supported by virtiofsd
+        1 << VIRTIO_F_VERSION_1 | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits() | 1 << 2
```
Yes you probably want the full set:

        1 << VIRTIO_F_VERSION_1
            | 1 << VIRTIO_F_NOTIFY_ON_EMPTY
            | 1 << VIRTIO_RING_F_INDIRECT_DESC
            | 1 << VIRTIO_RING_F_EVENT_IDX
            | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits()

drop the hardcoded 1 << 2

stsquad commented on Cargo.lock:
```diff
@@ -280,12 +551,12 @@ dependencies = [
 version = "0.1.0"
 dependencies = [
- "clap",
+ "clap 3.0.0-beta.2",
```
Any reason we need an explicit beta version? Comment?

Gaelan replied:
This is from vhost-user-i2c, which [explicitly requests that version](https://github.com/rust-vmm/vhost-device/blob/f42e35e72c5661bb65b3c38356c67d67b69058bd/i2c/Cargo.toml#L15) (presumably because it wants 3.0, but 3.0 wasn't released at the time).

vhost-user-scsi depends on clap v2 (via structopt), so adding it made Cargo rename vhost-device-i2c's clap dependency in the lockfile to disambiguate between the two versions in use.

vireshk replied:
We need to stick with this beta version for now as the next version needs a newer version of Rust (1.54 I think) and the container doesn't support that for now. It is going to be updated soon I heard. Once that is done, we can go to 3.0.7 then.

gsserge replied:
Hey, this repo is now using the latest rust-vmm-ci with Rust 1.54.

vireshk replied:
@Gaelan I moved to a non-beta version of clap now and it doesn't support the yaml stuff anymore, but it has a very simple interface available now. Please look at:

https://github.com/rust-vmm/vhost-device/pull/93

stsquad reviewed (changes requested):
Sorry about the noise - just getting used to the review tools. See comments inline

Gaelan commented:
I've pushed my changes here as new commits for now so it's easier to see what changed; we probably should squash them into the original commits before merging this.

vireshk commented:
> I've pushed my changes here as new commits for now so it's easier to see what changed; we probably should squash them into the original commits before merging this.

I haven't looked into the details yet, but lots of things have changed in the i2c crate few weeks back. Specially the testing part. I hope you can move to the versions of dependencies I2C is using and also use the format of I2C crate (just to make all this aligned together), like the ThisError trait, etc, and make sure most of the stuff is getting tested so the score doesn't come down by a huge number.

Gaelan commented:
Alright, rebased onto main and updated to new versions of deps. Still to do:

- [ ] make changes for consistency with `vhost-user-i2c` (ThisError, etc)
- [ ] add a few more tests to bring coverage in line
- [ ] support all feature flags (IIRC, some of the feature flags require handling on my end)
- [ ] figure out why @stsquad can't get it to run

Ablu commented:
I rebased this in https://github.com/rust-vmm/vhost-device/pull/301.
